src/learning/toolkit/cross_validation.rs

```diff
@@ -0,0 +1,235 @@
+        let test_targets = copy_rows(&targets, p.test_indices.clone());
+
+        model.train(&train_inputs, &train_targets);
+        let outputs = model.predict(&test_inputs);
+        costs.push(C::cost(&outputs, &test_targets));
+    }
+
+    costs
+}
+
+// TODO: Use a preallocated buffer for each fold.
+fn copy_rows<'a, I>(mat: &Matrix<f64>,
+                    rows: I) -> Matrix<f64>
+    where I: ExactSizeIterator<Item=&'a usize>
+{
+    let num_rows = rows.len();
+    let mut data = vec![0f64; num_rows * mat.cols()];
+    let mut idx = 0;
+    for &row in rows {
+        for col in 0..mat.cols(){
+            data[idx] = mat[[row, col]];
+            idx += 1;
+        }
+    }
+    Matrix::<f64>::new(num_rows, mat.cols(), data)
+}
+
+/// A permutation of 0..n.
+struct ShuffledIndices(Vec<usize>);
+
+/// Permute the indices of the inputs samples.
+fn create_shuffled_indices(num_samples: usize) -> ShuffledIndices {
+    let mut indices: Vec<usize> = (0..num_samples).collect();
+    in_place_fisher_yates(&mut indices);
+    ShuffledIndices(indices)
+}
+
+/// A partition of indices of all available samples into
+/// a training set and a test set.
+struct Partition<'a> {
+    train_indices: TrainingIndices<'a>,
+    test_indices: TestIndices<'a>
+}
+
+#[derive(Clone)]
+struct TestIndices<'a>(Iter<'a, usize>);
+
+#[derive(Clone)]
+struct TrainingIndices<'a> {
+    chain: Chain<Iter<'a, usize>, Iter<'a, usize>>,
+    size: usize
+}
+
+impl<'a> TestIndices<'a> {
+    fn new(indices: &'a [usize]) -> TestIndices<'a> {
+        TestIndices(indices.iter())
+    }
+}
+
+impl<'a> Iterator for TestIndices<'a> {
+    type Item = &'a usize;
+
+    fn next(&mut self) -> Option<&'a usize> {
+        self.0.next()
+    }
+}
+
+impl <'a> ExactSizeIterator for TestIndices<'a> {
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+impl<'a> TrainingIndices<'a> {
+    fn new(left: &'a [usize], right: &'a [usize]) -> TrainingIndices<'a> {
+        let chain = left.iter().chain(right.iter());
+        TrainingIndices {
+            chain: chain,
+            size: left.len() + right.len()
+        }
+    }
+}
+
+impl<'a> Iterator for TrainingIndices<'a> {
+    type Item = &'a usize;
+
+    fn next(&mut self) -> Option<&'a usize> {
+        self.chain.next()
+    }
+}
+
+impl <'a> ExactSizeIterator for TrainingIndices<'a> {
+    fn len(&self) -> usize {
+        self.size
+    }
+}
+
+/// An iterator over the sets of indices required for k-fold cross validation.
+struct Folds<'a> {
+    num_folds: usize,
+    fold_size: usize,
+    indices: &'a[usize],
+    count: usize
+}
+
+impl<'a> Folds<'a> {
+    fn new(indices: &'a ShuffledIndices, num_folds: usize) -> Folds<'a> {
+        let num_samples = indices.0.len();
+        assert!(num_folds > 1 && num_samples >= num_folds,
+            "Require num_folds > 1 && num_samples >= num_folds");
+        assert!(num_samples % num_folds == 0,
```

Comment:
I believe we'd prefer not to be bound by this assertion. For example this means that if we have a prime number of examples we can only use _leave-one-out_ cross validation.

The [KFold implementation in scikit-learn](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.KFold.html#sklearn.cross_validation.KFold) uses the following rule:

> The first n % n_folds folds have size n // n_folds + 1, other folds have size n // n_folds.

We can settle for the simpler implementation that exists now if this is a huge headache to implement.